src/cluster_legacy.c

```diff
@@ -6400,6 +6402,25 @@ void clusterCommandSetSlot(client *c) {
             clusterDoBeforeSleep(CLUSTER_TODO_SAVE_CONFIG | CLUSTER_TODO_UPDATE_STATE | CLUSTER_TODO_FSYNC_CONFIG);
         }
 
+        /* If we are a replica and executing the CLUSTER SETSLOT command from
+         * my primary, and my primary left without slots, we should turn into a
+         * replica of the new primary. */
+        if (nodeIsReplica(myself) && slot_was_my_primary && my_primary->numslots == 0 &&
+            server.cluster_allow_replica_migration) {
+            serverLog(LL_NOTICE,
+                      "Lost my last slot during slot migration. Reconfiguring myself "
+                      "as a replica of %.40s (%s) in shard %.40s",
+                      n->name, n->human_nodename, n->shard_id);
+            /* We are a replica and the client is actually my primary, the following
+             * clusterSetPrimary will free the client so we need to protect it. */
+            protectClient(c);
+            /* We are migrating to a different shard that has a completely different
+             * replication history, so a full sync is required. */
+            clusterSetPrimary(n, 1, 1);
+            unprotectClient(c);
+            clusterDoBeforeSleep(CLUSTER_TODO_SAVE_CONFIG | CLUSTER_TODO_UPDATE_STATE | CLUSTER_TODO_FSYNC_CONFIG);
+        }
+
```

Comment:
Maybe simplify this a bit? I think we can skip `unprotectClient`

```c
          clusterNode *my_primary = clusterNodeGetPrimary(myself);
          int slot_was_mine = server.cluster->slots[slot] == my_primary;
          clusterDelSlot(slot);
          clusterAddSlot(n, slot);

          /* If replica migration is allowed, check if the primary of this shard loses its last slot and the shard becomes empty. */
          if (server.cluster_allow_replica_migration && slot_was_mine && my_primary->numslots == 0) {
              serverAssert(n != my_primary);
              serverLog(LL_NOTICE,
                        "Lost my last slot during slot migration. Reconfiguring myself "
                        "as a replica of %.40s (%s) in shard %.40s",
                        n->name, n->human_nodename, n->shard_id);
              /* `c` is the primary client if `myself` is a replica, prevent it
               * from being freed by clusterSetPrimary. */
              if (nodeIsReplica(myself)) protectClient(c);
              /* We are migrating to a different shard that has a completely different
               * replication history, so a full sync is required. */
              clusterSetPrimary(n, 1, 1);
              clusterDoBeforeSleep(CLUSTER_TODO_SAVE_CONFIG | CLUSTER_TODO_UPDATE_STATE | CLUSTER_TODO_FSYNC_CONFIG);
          }
```

Reply:
yean, this work for me. My original idea was to make them their own blocks so they are easier to read, but it seems like there is too much code.